### DISCUSSION TITLE
Dealing with recursive constraints

### DISCUSSION THREAD

**Discussion by dabrahams** [hylo-lang/hylo#1216] [Language design]
Recursive protocol constraints cause an enormous amount of complexity in Swift.  This thread is about how to deal with that.  Obviously we could implement Swift's Requirement Machine, which is complex.  Maybe there are lighter-weight approaches, that start by banning recursive constraints.

Is this the minimal example that demonstrates a problematic recursive constraint?

```swift
trait Collection {
  type SubSequence: Collection
}
```
or do we need some other constraints on SubSequence in order to have a problem, e.g.

```swift
trait Collection {
  type SubSequence: Collection where
    SubSequence.SubSequence == SubSequence, // <=== THIS
    SubSequence.Element == Self.Element,
    SubSequence.Position == Self.Position
}
```

…which is more like  Swift's `SubSequence` constraints.  In Swift, the `SubSequence` type is the result of the slicing operation on a collection.

If we want to ban recursive constraints, and the first example is a minimal problematic example, practically speaking, it means that `SubSequence` could not be an associated type.  Well, we could make it an associated type with weaker constraints, but I lived in that world for a long time in Swift and the consequences were really ugly.

So that means slicing a `C: Collection` must yield some non-associated type; let's call it `Slice<C>`:

```swift
type Slice<Base: Collection>: Collection {
   remote var base: Base
   var startPosition: Base.Position
   var endPosition: Base.Position
   ...
}
```

If we just have the above declaration, here's what Swift allows that Hylo would not:
1. I can create a more-efficient custom `SubSequence` type for a given `Collection` type.  For example, slicing a `Range<Int>` yields a `Range<Int>` instead of something that stores four `Int`s.  `String` has a custom `SubSequence` with a smaller start/end index representation when the sliced `String` is stored inline.
2. I can ensure that we never end up creating `Slice<Slice<Array<Int>>>`, because slicing a `Slice<T>` yields `Slice<T>`.
3. I can nail down the requirement that any custom `SubSequence` type `X` also has the property that `X.SubSequence == X`.

I consider the last of these completely dispensable.  My next question would be, how much of the other capabilities can we claw back for Hylo?

**Comment by dabrahams** [comment]
In the interest of completeness, I'll mention that we could get some of the capability back by creating an overloaded slicing subscript on `Slice<Base>` that returns `Self`, and teaching overload resolution that it is to be preferred over the more general slicing subscript of `Collection`.  I personally find that highly unsatisfying from a usability point of view.  Overloading in general is bad for users, increasing the number of distinct APIs that must be understood in order to grasp what is conceptually one operation.  This is overloading on return type, which is even worse in many ways.  I consider that solution to be worse than the problems it solves.

**Reply by dabrahams** [reply]
The “make `SubSequence` an associated type with weaker constraints” approach taken by pre-recursive-constraint-capable Swift was ugly primarily because then your slice was not a `Collection` in generic contexts, so it didn't interoperate with the rest of the library the way it should; for example, none of the `Collection`-requiring algorithms could work on it.

But a variation on that theme might work.  “Temporary limitations” of the current Hylo compiler prevent me from type checking this except in my head, but it plays out something like the following (I could probably prove it out using Swift, if that's helpful).  Contract documentation intentionally omitted so you can more easily see the type system mechanics.

```swift
/// Everything but the slicing-related requirements of Collection
trait CollectionCore {

  type Element
  type Index

  fun start_index() -> Index
  fun end_index() -> Index
  fun index(after i: Index) -> Index

  subscript(_ i: Index): Element { let }

  // A type that is wrapped to get the result of slicing
  type SliceCore
    = DefaultSliceCore<Self> // Default makes new Collections easier to define.

  // Not for user consumption; provides the implementation of slicing
  subscript slice_core(_ start: Index, _ end: Index): SliceCore { let }
}

// Adds slicing support requirements to CollectionCore.
trait Collection: CollectionCore {

  type SliceCore: CollectionCore where
    SliceCore.SliceCore == SliceCore

  // `Unsliced`: `Self == `Slice<U> ? U : Self`
  type Unsliced: CollectionCore
    where Unsliced.SliceCore == SliceCore,
          Unsliced.Element == SliceCore.Element,
          Unsliced.Index == SliceCore.Index

}

// Provide the user-consumed slicing operation.
extension Collection {

  typealias Unsliced = Self

  subscript(_ start: Index, _ end: Index): Slice<Unsliced> {
    yield Slice(core: self.sliceCore[start, end])
  }

}

// Wraps Unsliced_'s SliceCore as a full Collection.
type Slice<
  Unsliced_: CollectionCore
  where Unsliced_.SliceCore: CollectionCore,
        Unsliced_.SliceCore.SliceCore == Unsliced_.SliceCore,
        Unsliced_.SliceCore.Index == Unsliced_.Index,
        Unsliced_.SliceCore.Element == Unsliced_.Element
>: Collection
{

  memberwise init

  var core: SliceCore

  typealias Unsliced = Unsliced_

  // Prevents buildup of Slice<...> nesting.
  typealias SliceCore = Unsliced.SliceCore

  // Forward CollectionCore requirements to core
  typealias Element = SliceCore.Element
  typealias Index = SliceCore.Index
  fun start_index() -> Index { core.start_index() }
  fun end_index() -> Index { core.end_index() }
  fun index(after i: Index) -> Index {  core.index(after: i) }
  subscript(_ i: Index): Element { _read { yield core[i] } }

  subscript slice_core(_ start: Index, _ end: Index): SliceCore {
    yield core.slice_core[start, end]
  }

}

conformance Slice: Equatable where Unsliced.SliceCore: Equatable {}

//
// Conveniences so new Collections don't need to define their own slicing support.
//

extension Collection where SliceCore == DefaultSliceCore<Self> {

  subscript slice_core(_ start: Index, _ end: Index): SliceCore {
    yield DefaultSliceCore(base: self, start: start, end: end)
  }

}

type DefaultSliceCore<Base: CollectionCore>: CollectionCore  {

  let base: remote let Base
  var start: Index
  var end: Index

  typealias Element = Base.Element
  typealias Index = Base.Index

  fun start_index() -> Index { start }
  fun end_index() -> Index { end }
  fun index(after i: Index) -> Index { base.index(after: i) }

  subscript(_ i: Index): Element { yield base[i] }

  subscript slice_core(_ start: Index, _ end: Index): Self {
    yield Self(base: self.base, start: start, end: end)
  }

}

conformance DefaultSliceCore: Equatable where Base: Equatable, Index: Equatable {}

// ========= Demonstrations ============

// A simple collection using default slicing
type Basic: Collection, Equatable {

  typealias Element = String
  typealias Index = Int

  fun start_index() -> Index { 0 }
  fun end_index() -> Index { 10 }
  fun index(after i: Index) -> Index { i + 1 }

  subscript(_ i: Index): Element { yield String(i) }

}

 fun test_basic() {
  let basic = Basic()

  precondition(
    basic[3, 5] ==
      Slice<Basic>(
        core: DefaultSliceCore(base: basic, start: 3, end: 5)))

  precondition(basic[3, 5][3, 5] == basic[3, 5])
}

// A collection with a custom slice representation.
type CustomSliced: Collection, Equatable {

  typealias Element = UInt8
  typealias Index = Int

  fun start_index() -> Index { 0 }
  fun end_index() -> Index { 10 }
  fun index(after i: Index) -> Index { i + 1 }

  subscript(_ i: Index): Element { yield UInt8(i) }
  subscript slice_core(_ start: Index, _ end: Index): SliceCore {
    yield SliceCore(start: start, end: end)
  }

  type SliceCore: Collection, Equatable {

    typealias SliceCore = Self
    typealias Element = UInt8
    typealias Index = Int

    var start: Int
    var end: Int

    fun start_index() -> Index { start }
    fun end_index() -> Index { end }
    fun index(after i: Index) -> Index { i + 1 }

    subscript(_ i: Index): Element { yield UInt8(i) }

    subscript slice_core(_ start: Index, _ end: Index): SliceCore {
      yield Self(start: start, end: end)
    }
  }

}

fun test_custom() {
  let custom = CustomSliced()

  precondition(
    custom[3, 5] ==
      Slice<CustomSliced>(core: CustomSliced.SliceCore(start: 3, end: 5)))

  precondition(custom[3, 5][3, 5] == custom[3, 5])
}

fun main() {
  test_basic()
  test_custom()
}
```

**Reply by dabrahams** [reply]
OK, translating to Swift and running it through the compiler found lots of silly errors, but didn't change the basic shape of things.  The result is [here](https://gist.github.com/dabrahams/bc9497acdc20001cde14ed57bc4b4965) in two files, the second being a refined version of the first.  I've translated the second refinement into Hylo above.

**Comment by dabrahams** [comment]
If the above technique works, we'll want to think about how to hide from the user the kludgey parts that are only there for the sake of the technique, such as `CollectionCore`.  I have some ideas.

**Comment by dabrahams** [comment]
@slavapestov points out this Swift example, which illustrates another source of undecidability that doesn't involve recursive constraints.  That strongly suggests to me that undecidability is inevitable, and the best we can do is to put computational limits on typechecking.  He also points out that the requirements machine can be implemented in a few hundred lines of code so maybe we should just bite the bullet.

```swift
protocol MyIterator {
  associatedtype Element
}

protocol MySequence {
  associatedtype Iterator: MyIterator
}

struct ConcreteIterator<Elements: MySequence>: MyIterator {
  typealias Element = Elements.Iterator.Element
}
```

The definition of ConcreteIterator creates an implicit inequality constraint, “where Elements.Iterator != Self” because clearly `Elements.Iterator == Self` makes `ConcreteIterator.Element` a circular definition.  However, detecting all such inequality constraints is equivalent to solving the halting problem.

**Reply by kyouko-taiga** [reply]
Perhaps you should show the illegal declaration to better understand the problem:
```swift
struct ConcreteSequence: MySequence {
  typealias Iterator = ConcreteIterator<Self>
}
```

But I'm not sure this particular case is proving undecidability. IIUC, Slava's question was not about determining whether or not the program was ill-formed, but about assigning blame. IMO the illegal line is the type alias in `ConcreteSequence` and I think we can decide that it is ill-formed if we try to enumerate the consequences of the declaration and proving them one after another.

```
typealias Element = Elements.Iterator.Element
  => ConcreteSequence.Iterator == ConcreteIterator<ConcreteSequence>
  => ConcreteIterator<ConcreteSequence> : MyIterator
  => ConcreteIterator<ConcreteSequence>.Element == ConcreteSequence.Iterator.Element
  => ConcreteIterator<ConcreteSequence>.Element == ConcreteIterator<ConcreteSequence>.Element
```

- The first equality constraint is caused by the type alias in `ConcreteSequence`. It type checks because it can't contradict anything yet. So we can take it for granted and move on.
- The conformance constraint derives from the fact that `MySequence.Iterator` must model `MyIterator`. Conformance constraints always check, so we're still good. We can move on.
- The next equality constraint is consequence of the conformance. We find it by looking at how `ConcreteIterator<ConcreteSequence>` implements `MyIterator.Element`. Here we can realize that the definition is circular because we already decided that `ConcreteSequence.Iterator` was `ConcreteIterator<ConcreteSequence>`.

This algorithm relies on the ability to have a normal form for all associated types, which I believe is possible without recursive constraints. It may be naive but I really think that if we abandon the idea that all constraints everywhere must be proven "at the same time", then we can simplify our problem. At the very least we can assign blame more easily.

**Reply by dabrahams** [reply]
I disagree about the source of the illegality.  What I showed is sufficient.  `ConcreteIterator` is missing an inequality constraint.

**Reply by kyouko-taiga** [reply]
I don't think so. We'll have to take this online for me to understand your position.

**Reply by dabrahams** [reply]
I think it's very simple and I'll try to explain it here; if that fails of course I'm happy to discuss in another venue.

For separate checking of generics to be complete, a generic component must describe, in the `<...>` and the `where` clause, all constraints that its parameters must satisfy in order for a concretization of that component to pass typechecking.  When `X.Iterator == ConcreteIterator<X>`, which is allowed by the current constraints on `ConcreteIterator`, you get this:

```swift
struct ConcreteIterator_X: MyIterator {
  typealias Element = ConcreteIterator_X.Element
}
```
 
This definition can't pass typechecking because `Element` is a circular definition, therefore there's a constraint missing.

```swift
struct ConcreteIterator<Elements: MySequence>: MyIterator 
  where Elements.Iterator != Self // <== HERE
{
  typealias Element = Elements.Iterator.Element
}
```

`ConcreteSequence.Iterator` doesn't satisfy this constraint and would be rejected if it were present.

Saying that `ConcreteIterator` does _not_ have an error in its definition requires giving up on the idea that separate checking is complete.  I haven't worked out the example yet, but I'm pretty sure we can show cases where having this incompleteness is arbitrarily bad and would necessitate the equivalent of a template instantiation backtrace in diagnostics.

**Reply by kyouko-taiga** [reply]
Sorry, I'm still not on board. If the inequality can be deduced from the specialization, then all is well and having it spelled out explicitly would only be a hint for the type checker. If we had to ensure that the equality constraints of generic signature can never be contradicted by a given set of arguments, then we'd have to say that `<T: Collection where T.Element == Int>` is underspecified because we must also add `T != Array<Bool>, T != Deque<Bool>, T != LinkedList<Bool>, ...`.

Separate checking is not affected AFAICT because there are still no possible specialization of your `ConcreteIterator`'s _declaration_ that can violate its _implementation_. The latter can be checked once and for all, and then compiled using existentialization. That is unlike template specialization that can always fail because of an underspecified signature:

```c++
template<typename T: AdditiveArithmetic>
T foo(T a, T b) { return a * b; }
```

Here, because `T.infix*` doesn't exist in `AdditiveArithmetic`, we can't type check `foo`'s implementation before we receive its generic arguments.

**Reply by dabrahams** [reply]
> Separate checking is not affected AFAICT because there are still no possible specialization of your ConcreteIterator's _declaration_ that can violate its _implementation_. The

Which parts do you consider declaration and which do you consider implementation?

>  If the inequality can be deduced from the specialization, then all is well and having it spelled out explicitly would only be a hint for the type checker

How is that different from this:

```swift
func f<T: Equatable, U>(x: T, y: U) -> Bool { x == y }
```

I could say all is well because for any given specialization of `f`, we can deduce that `T == U` and spelling it out explicitly would only be a hint for the type checker.  I don't consider the bodies of generic types to be any different from the bodies of generic functions where separate checking is concerned.  Once the generic type is typechecked, no arguments satisfying the constraints should cause an error.  

Now, I suppose one could argue that `ConcreteSequence: MySequence` is the error and thus `ConcreteSequence` doesn't satisfy the constraints of `MyIterator`.  But I can't explain _why_ `ConcreteSequence` doesn't conform to `MySequence` without referring into the body of `ConcreteIterator`, which already passed typechecking in your world. 

I'm still trying to come up with the example that demonstrates what I mean more clearly in Swift but at the moment I suspect it might require HKTs (or parameterized associated types).  Although I wouldn't be totally shocked if scoped conformances also shows it.

**Reply by dabrahams** [reply]
Here's an example using HKTs (the best I could do for now).  I claim there's some constraint missing on `shouldNotTypeCheck` that should prevent `MakeIterator.IteratorType<T> == ConcreteIterator<T>`, and that constraint would be derived from the missing constraint on `ConcreteIterator`.

```swift
// This is a sort of higher-kinded type.
protocol IteratorSupplier {
  associatedtype IteratorType<S: MySequence>: MyIterator
}

struct MakeConcreteIterator: IteratorSupplier {
  typealias IteratorType<S: MySequence> = ConcreteIterator<S>
}

func shoudlNotTypeCheck<MakeIterator: IteratorSupplier>(_: MakeIterator.Type) {

  struct ConcreteSequence: MySequence {
    typealias Iterator = MakeIterator.IteratorType<Self>
  }

  print(ConcreteSequence.Iterator.Element.self) // Would crash Swift
}

shouldNotTypeCheck(MakeConcreteIterator.self)
```

I realize we haven't bought into HKTs but there are other restrictions in Swift's generic system that we intend to lift that I believe could provide similar expressivity, and thus problems.

**Reply by kyouko-taiga** [reply]
> How is that different from this:
>
> ```
> func f<T: Equatable, U>(x: T, y: U) -> Bool { x == y }
> ```

Clearly it doesn't type check because there is nothing in the signature that says `T.infix==` is defined for an argument of type `U`. The important property that separate checking upholds is precisely that it can reject this function. Template specialization can't.

It may be worth making sure we agree on some ground principles, so let me try to define them.

Inside of `f`, `T` and `U` are bound existentially and therefore only described by their constraints. So we take these constraints for granted and operate under a closed-world assumption. When we call `f`, we see universally quantified variables so we're free to pick anything as long as we can prove that our choices satisfy `f`'s signature.

There is no way to skip this proof, even without equality constraints. If we write this:
```
func g<T: Equatable>(x: T, y: T) -> Bool { x == y }
```
We sill have to prove that whatever we substitute for `T` is indeed equatable when we call `g`. So at call site we must first substitute generic parameters and _then_ check that all constraints on them are satisfied. That is fine because the existentially quantified version was rich enough to be checked separately.

Now let's discuss about the constraints.

> For separate checking of generics to be complete, a generic component must describe, in the <...> and the where clause, all constraints that its parameters must satisfy in order for a concretization of that component to pass typechecking.

This definition is too strong to support anything other than conformance constraints, and only in a language where `T: P` can never rule out `T: Q`. As I mentioned, `<T: Collection where T.Element == Int>` is underspecified under this definition because it doesn't tell about all the types that `T` cannot be. There is an unbounded number of such types so we can't even write a complete signature.

And if we're ready to say that `T == Int` implies that `T != Bool`, then I don't say why we can't also say that the generic signature of `ConcreteIterator` implies `Elements.Iterator != Self`. I don't think making such an inference is undecidable and I'm pretty sure Slava didn't say it was.

Now even if I'm wrong and making these inferences is indeed undecidable, then the cure is just to run type checking under a time budget. I never understood why type checking _had_ to be decidable and I thought we agreed to give up on this idea a long time ago.

> Here's an example using HKTs [...]

`IteratorSupplier` is not really higher-kinded but that is not the point of your example. Basically you're trying to hide the implicit constraints on `MakeIterator` that are caused by the definition of `ConcreteSequence`. There's a simpler way to do that:

```swift
func h<T>(_ x: T) {
  struct S<U: Collection> where U.Element: Equatable, U.Element == T {}
}
```

Here, `<T>` is underspecified because it doesn't say that `T` must be equatable.

*Note: This function is rejected in Swift because we can't declare a generic type in a generic function but we can replace `S` by a generic function with the same signature, in which case it may be a source of unsoundness worth invistigating.*

This problem is slightly orthogonal IMO. At the very least there are many more considerations to take into account when it comes to declarations nested in generics so I think that would deserve a separate design discussion cycle.

**Reply by dabrahams** [reply]
You never answered this question, which may in fact play a 🔑  role in understanding:

> Which parts do you consider declaration and which do you consider implementation?

>> How is that different from…?
>
> Clearly it doesn't type check because there is nothing in the signature that says T.infix== is defined for an argument of type U. 

In the same way there's nothing in the signature of `ConcreteIterator` that says `Elements.Iterator != ConcreteIterator<Elements>`, and yet that condition produces a type error in the body 🔑 of `ConcreteIterator`.  You haven't said anything that explains why these cases should be viewed as different.

> The important property that separate checking upholds is precisely that it can reject this function. Template specialization can't.

I don't know why you're bringing up template specialization; that's not how I've been thinking of this.  Furthermore, to the extent that what I'm talking about is anything like what happens with templates, it's like specializing the template with maximally-uncooperative archetypes, and that technique certainly does reject `f` in my example.  But please, let's not bring templates into this.  The ground principles you outline are all part of my background assumption, except possibly for this, which I don't understand:

> we must first substitute generic parameters and _then_ check that all constraints on them are satisfied

What specifically do you mean by "substitute generic prarameters" (e.g. substitute them for what?) and why would it be important to do that before checking that the generic arguments satisfy constraints on the corresponding generic parameters?  (speculating: maybe this “substitution” is just the way you're talking about identifying what the generic arguments are?).  And why are you putting so much emphasis on this ordering?

>> For separate checking of generics to be complete, a generic component must describe, in the <...> and the where clause, all constraints that its parameters must satisfy in order for a concretization of that component to pass typechecking.
>
> This definition is too strong to support anything other than conformance constraints, and only in a language where T: P can never rule out T: Q. 

I don't understand what it could mean for my definition of “complete checking” to “support” a language feature; the only way I can figure out to read the first part of your sentence (which may be a misinterpretation) is "that kind of completeness can't be provided by any language with non-conformance (e.g. type equality) constraints.”  But that's clearly false. And I can't even guess at an interpretation for the second part of your sentence, I'm afraid.

> As I mentioned, `<T: Collection where T.Element == Int>` is underspecified…

I didn't address this before so let me go back to the original statement:

> If we had to ensure that the equality constraints of generic signature can never be contradicted by a given set of arguments, then we'd have to say that `<T: Collection where T.Element == Int>` is underspecified because we must also add `T != Array<Bool>`, `T != Deque<Bool>`, `T != LinkedList<Bool>`, ....

I don't understand what the premise “if we had to ensure that the equality constraints of generic signature can never be contradicted by a given set of arguments” even means, but it doesn't sound like anything I'm talking about.  `<T: Collection where T.Element == Int>` already rules out `T` being a collection whose `Element` is `Bool`.  This example is also very unlike `ConcreteIterator` because I'm saying the _body_ 🔑 of `ConcreteIterator` implies constraints on its parameters that are unstated.  There is no body in what you're describing.

> if we're ready to say that `T == Int` implies that `T != Bool`, then I don't say why we can't also say that the generic signature of `ConcreteIterator` implies `Elements.Iterator != Self`.

Why we can't also say that is very simple: I can change the _body_ of `ConcreteIterator` (which is not part of its generic signature) in such a way that you would never suggest such an implication exists

```swift
struct ConcreteIterator<Elements: MySequence>: MyIterator {
  typealias Element = Int // <==
}
```
Poof! Same generic signature 🔑, no constraint on the relationship between `Elements.Iterator` and `Self`.

> I don't think making such an inference is undecidable and I'm pretty sure Slava didn't say it was.

Nor did I say that.  My understanding of all the undecidability we've seen is that it is something that _can_ occur given a certain level of expressivity.  The simplest known example that actually demonstrates it for the problem @slavapestov originally pointed out (years ago now) is much more convoluted than what we've been looking at. What I believe Slava said is that there _exists_ a Swift program where the existence-or-not of what I'm calling an implied inequality constraint is undecidable.  Nobody has produced an example, but I believe Slava arrived at his conclusion by mapping the problem of determining the existence of such a constraint onto the problem of detecting infinite recursion (i.e. non-termination) in a functional program that you could derive mechanically from the declarations in the problematic Swift program.

> you're trying to hide the implicit constraints on `MakeIterator` that are caused by the definition of `ConcreteSequence`

Are you sure? 🤷 That doesn't sound like anything I was trying to do.  I was trying to demonstrate that those implicit **inequality** constraints exist and need to be expressed (e.g. on `shouldNotTypeCheck`), and can't be inferred by the compiler from declarations _even if it is willing to look into the bodies of generic types_ to find them.  Even if you think the body of a generic type is part of its generic signature 🔑 , I'm pretty sure you don't think the body of a generic function is part of _its_ signature; I _think_ we agree that the compiler should not look into the body of a (generic) function in order to determine that a call to that function type-checks.

Your "simpler way" doesn't demonstrate the same thing I was trying to show because there's no implied **inequality** constraint.

**Reply by kyouko-taiga** [reply]
> You never answered this question, which may in fact play a 🔑 role in understanding:

> > Which parts do you consider declaration and which do you consider implementation?

A type isn't like a function in that it has public parts in its body. In particular, the parts that make it conform to a trait must be visible to every consumer of that conformance. A function exports nothing more than it's signature.

I think it is the point where asynchronous messages on the Internet have reached the maximum they can do for us. I carefully read everything you wrote and would prefer to continue online.

**Reply by dabrahams** [reply]
I'm ready to have an interactive discussion whenever you are.  But FWIW, I put a lot of labor into the post above. I asked very specific questions where I could in order to try to understand what you're saying (and you still haven't even answered the first one, although you seem to be sort of hinting at an answer).  Where I couldn't come up with specific questions I tried hard to explain why I couldn't understand you and to point out mistaken assumptions you seemed to be making about what I've said.  It's going to be pretty unsatisfying for me if we don't address all of those points in whatever discussion comes next.  And I think it's important that we ultimately tie up all the loose ends here so that the written record can act as a guide for others, and for our future selves.

**Reply by kyouko-taiga** [reply]
I'm hoping we can get to the very bottom of everything and it'll be very valuable to come back here with our conclusions. I just feel like continuing the discussion in this form at this point is a time sink for both of us with endless opportunities to get misunderstood.

**Comment by dabrahams** [comment]
FWIW, this is what I get when I do as you suggested and replace S by a generic function with the same signature 

```swift
func h<T>(_ x: T) {
  func s<U: Collection>(_: U.Type) where U.Element: Equatable, U.Element == T {}
}
```

The compiler accepts it, which seems perfectly fine to me.  If you add the call `s([x])` you get an error, as you should because `T` is not `Equatable`; in fact there are no types you can call `s` with.  This program (or your original version) should only be rejected if you believe declaring a generic whose constraints can't be satisfied must be diagnosed by the compiler.  To me such a diagnosis seems like a nice-to-have, but non-critical, feature.  I certainly can't see any hint of unsoundness in its absence.
